From 31883f21eff4265b68bb36f67b254adb524db6ae Mon Sep 17 00:00:00 2001
From: guoxiaoqi <guoxiaoqi2@huawei.com>
Date: Thu, 16 Apr 2020 14:51:44 +0800
Subject: [PATCH] bugfix-openssh-6.6p1-log-usepam-no

---
 sshd.c      | 4 ++--
 sshd_config | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/sshd.c b/sshd.c
index c6c03ae..c291a5e 100644
--- a/sshd.c
+++ b/sshd.c
@@ -1812,9 +1812,9 @@ main(int ac, char **av)
 	parse_server_config(&options, rexeced_flag ? "rexec" : config_file_name,
 	    cfg, &includes, NULL);
 
-	/* 'UsePAM no' is not supported in Fedora */
+	/* 'UsePAM no' is not supported in openEuler */
 	if (! options.use_pam)
-		logit("WARNING: 'UsePAM no' is not supported in Fedora and may cause several problems.");
+		logit("WARNING: 'UsePAM no' is not supported in openEuler and may cause several problems.");
 
 #ifdef WITH_OPENSSL
 	if (options.moduli_file != NULL)
diff --git a/sshd_config b/sshd_config
index e125992..ebc28b3 100644
--- a/sshd_config
+++ b/sshd_config
@@ -87,7 +87,7 @@ AuthorizedKeysFile	.ssh/authorized_keys
 # If you just want the PAM account and session checks to run without
 # PAM authentication, then enable this but set PasswordAuthentication
 # and KbdInteractiveAuthentication to 'no'.
-# WARNING: 'UsePAM no' is not supported in Fedora and may cause several
+# WARNING: 'UsePAM no' is not supported in openEuler and may cause several
 # problems.
 #UsePAM no
 
-- 
2.23.0

