From 41c115a5ea1cb79a6a3182773c58a23f760e8076 Mon Sep 17 00:00:00 2001
From: Damien Miller <djm@mindrot.org>
Date: Wed, 12 Sep 2018 16:50:01 +1000
Subject: [PATCH 016/294] delete the correct thing; kexfuzz binary

---
 Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.in b/Makefile.in
index d3cffc5..46562ad 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -306,7 +306,7 @@ distclean:	regressclean
 	rm -f regress/unittests/pkcs11/*.o
 	rm -f regress/unittests/pkcs11/test_pkcs11
 	rm -f regress/misc/kexfuzz/*.o
-	rm -f regress/misc/kexfuzz
+	rm -f regress/misc/kexfuzz/kexfuzz$(EXEEXT)
 	(cd openbsd-compat && $(MAKE) distclean)
 	if test -d pkg ; then \
 		rm -fr pkg ; \
-- 
1.8.3.1

